pkg/kubelet/cm/cpumanager/policy_static.go

```diff
@@ -348,13 +349,93 @@ func (p *staticPolicy) GetTopologyHints(s state.State, pod *v1.Pod, container *v
 
 	// Generate hints.
 	cpuHints := p.generateCPUTopologyHints(available, reusable, requested)
-	klog.Infof("[cpumanager] TopologyHints generated for pod '%v', container '%v': %v", pod.Name, container.Name, cpuHints)
+	klog.Infof("[cpumanager] TopologyHints generated for pod '%v', container '%v': %v", format.Pod(pod), container.Name, cpuHints)
 
 	return map[string][]topologymanager.TopologyHint{
 		string(v1.ResourceCPU): cpuHints,
 	}
 }
 
+func (p *staticPolicy) getPodRequestedCPU(pod *v1.Pod) int {
+	// The maximum of requested CPUs by init containers.
+	requestedByInitContainers := 0
+	for _, container := range pod.Spec.InitContainers {
+		if _, ok := container.Resources.Requests[v1.ResourceCPU]; !ok {
+			continue
+		}
+		requestedCPU := p.guaranteedCPUs(pod, &container)
+		if requestedCPU > requestedByInitContainers {
+			requestedByInitContainers = requestedCPU
+		}
+	}
+	// The sum of requested CPUs by app containers.
+	requestedByAppContainers := 0
+	for _, container := range pod.Spec.Containers {
+		if _, ok := container.Resources.Requests[v1.ResourceCPU]; !ok {
+			continue
+		}
+		requestedByAppContainers += p.guaranteedCPUs(pod, &container)
+	}
+
+	if requestedByInitContainers > requestedByAppContainers {
+		return requestedByInitContainers
+	}
+	return requestedByAppContainers
+}
+
+func (p *staticPolicy) GetPodTopologyHints(s state.State, pod *v1.Pod) map[string][]topologymanager.TopologyHint {
+	// Get a count of how many guaranteed CPUs have been requested by Pod.
+	requestedByPod := p.getPodRequestedCPU(pod)
+
+	// Number of required CPUs is not integer or a pod hasn't got a Guaranteed QoS class.
+	// It will be treated by Topology Manager as having no preference and cause to ignore this
+	// resource while considering a pod alignment.
+	// In definition of hints this is equal to: TopologyHints[NUMANodeAffinity: nil, Preferred: true].
+	if requestedByPod == 0 {
+		return nil
+	}
+
+	assignedCPUs := cpuset.NewCPUSet()
+	for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+		requested := p.guaranteedCPUs(pod, &container)
+		if allocated, exists := s.GetCPUSet(string(pod.UID), container.Name); exists {
+			if allocated.Size() != requested {
```

Comment:
Can you add the same comment we have in `GetTopologyHints()` above this, so that people looking at this function in isolation understand what this loop is trying to accomplish. Along with that, include a note on what `assignedCPUs` is tracking (i.e. the set of CPUs already assigned to containers in this pod).

Reply:
It doesn't look like this was done.